faker-js/faker
# fix: tree-shaking

# Summary

This PR addresses tree-shaking issues by optimizing the handling of locale imports and exports. The changes ensure that only the necessary locale modules are included in the final bundle, enhancing performance and reducing unnecessary file sizes.

This is a major requirement for the release of v9.0.

# Details

The previous implementation was causing unnecessary modules to be included during tree-shaking due to the additional export of the implicit `allLocales` variable. 

https://github.com/faker-js/faker/blob/03ac47198313126c3cc8cb3fb9ef184360bb79eb/src/index.ts#L39

This was refactored into an explicit export under `./locales/index.ts`. The `generate-locales` script has been updated to accommodate this change. 

Additionally, the package.json file has been updated to specify `"sideEffects": false`, which informs bundlers that all files within this package are side-effect-free and can be safely tree-shaken.

_To be perfectly honest, even after investigating this issue for multiple hours, I'm still not perfectly sure WHY this exact change fixes the issue._

**Please note, that only the combination of these changes will take any action. Applying one or the other change separately does not resolve the issue of everything being included in a final project bundle.**

# Testing

To verify these changes, follow these steps:

1. Check out this pull request locally.
1. Build the faker library using `pnpm run build`.
1. Clone our [playground repository](https://github.com/faker-js/playground).
1. Follow the instructions in the README.md file of the playground repository.
1. Override the faker dependency in the playground's `package.json` (replace "overrides-for-dev" with "overrides-for-dev").
1. Build/compile the playground projects `webpack-cjs` and `webpack-esm`.
1. Ensure that the console prints a bundle size number of around 400-480 KB when linking binaries from this pull request branch.
1. Compare this bundle size with the approximately 2.2 MB bundle size when linking binaries from the next branch.

These steps will confirm that the tree-shaking optimizations have reduced the bundle size appropriately and that the library functions as expected with the updated locale handling.

# Additional

I consider this a breaking change as of https://github.com/faker-js/faker/issues/2335#issuecomment-1687709538.

Fixes #1791, #2335.

- #1791 
- #2335
